File: sudoku_solver/sudoku.py
import time

board = [
    [0,0,0,0,0,0,0,0,1],
    [0,0,0,0,1,0,2,3,0],
    [0,0,8,4,3,0,0,0,7],
    [0,9,6,0,2,7,0,4,8],
    [0,0,0,0,0,9,0,0,0],
    [0,5,0,0,0,0,0,0,0],
    [0,0,4,0,0,5,0,0,0],
    [0,0,0,0,9,0,1,0,0],
    [9,8,2,7,0,0,0,0,0]
]

#board output
def output_sudoku(sudoku):
    for i in range(len(sudoku)):
        if i % 3 == 0 :
            print(" - - - - - - - - - - - - - - ")
        for j in range(len(sudoku[1])):
            if j % 3 == 0:
                print(" | ",end="")
            if j == 8:
                print(str(sudoku[i][j])+" | ")
            else:
                print(str(sudoku[i][j]) + " ",end="")

#checking whether the value in x,y fulfills the sudoku rules
def check_rules(x,y,sudoku):
    value=sudoku[y][x]
    #checking if the value is already in the same row or column
    for i in range(len(sudoku)):
        if sudoku[y][i] == value and i != x:
            return False
        if sudoku[i][x] == value and i != y:
            return False
    #checking if the value is in the same box
    #finding the box
    xmin = x // 3
    ymin = y // 3
    for i in range(ymin * 3, ymin * 3 + 3):
        for j in range(xmin * 3, xmin * 3 + 3):
            if sudoku[i][j] == value and i != y and j != x:
                return False
    return True

#if empty entry found, return the coordinates, if not return (-1,-1)
#just have to look at the entries for j>=y
def find_empty(y,sudoku):
    for j in range(y, len(sudoku)):
        for i in range(len(sudoku)):
            if sudoku[j][i] == 0:
                return (i,j)
    return (-1,-1)

#the backtracking solver
def backtracking_solver(sudoku):
    #instead of recursion, I used a stack
    stack = []
    #starting point (x,y)-coordinate
    pos=(0,0)
    pos = find_empty(pos[1],sudoku)
    #until there are zeros in the array
    while pos[0] != -1:
        #since it makes no sense to begin at 0
        sudoku[pos[1]][pos[0]] = 1
        value = 1
        #backtracking algo
        while not check_rules(pos[0],pos[1],sudoku):
            #if value == 9 and still no solution found, we want to go back in the list
            if value == 9:
                sudoku[pos[1]][pos[0]] = 0
                pos = stack.pop()
                #increase the popped value (backtracking algo)
                value = sudoku[pos[1]][pos[0]] + 1
                sudoku[pos[1]][pos[0]] = value
                #special case, if we grab a 9, we want to go back further
                if value > 9:
                    sudoku[pos[1]][pos[0]] = 0
                    pos = stack.pop()
                    value = sudoku[pos[1]][pos[0]] + 1
                    sudoku[pos[1]][pos[0]] = value
            else:
                #iterate from 1-8
                sudoku[pos[1]][pos[0]] += 1
                value = sudoku[pos[1]][pos[0]]
        #append the entry and look for the next zero
        stack.append(pos)
        pos = find_empty(pos[1],sudoku)

start_time = time.time()
backtracking_solver(board)
output_sudoku(board)
print("\n\n%s seconds" % (time.time() - start_time))

File: sudoku_solver/sudoku.cpp
#include <iostream>
#include <vector>
#include <stack>


//outputting the board
void out_sudoku(const std::vector<std::vector<short>>& sudoku)
{
    for (auto i = sudoku.begin(); i != sudoku.end(); ++i)
    {
        if (std::distance(sudoku.begin(), i) % 3 == 0)
            std::cout << " - - - - - - - - - - - - - - " << std::endl;

        for (auto j = i->begin(); j != i->end(); ++j)
        {
            if (std::distance(i->begin(), j) % 3 == 0)
            {
                std::cout << " | ";
            }
            if (std::distance(i->begin(), j) == 8)
            {
                std::cout << *j << " | " << std::endl;
            }
            else
            {
                std::cout << *j << " ";
            }
        }
    }
    std::cout << " - - - - - - - - - - - - - - " << std::endl; 

}

//checks whether the new value at the position pair fullfills the sudoku rules
bool check_rules(const std::pair<short, short>& pos, const std::vector<std::vector<short>>& sudoku) 
{
    short value = sudoku[pos.second][pos.first];
    //checking rows and columns
    for (short i = 0; i < sudoku.size(); i++) 
    {
        if (sudoku[pos.second][i] == value && i != pos.first)
        {
            return false;
        }
        if (sudoku[i][pos.first] == value && i != pos.second)
        {
            return false;
        }
    }
    //checking the ''square''
    short xmin = pos.first / 3;
    short ymin = pos.second / 3;
    for (short i = ymin * 3; i < ymin * 3 + 3; i++)
        for (short j = xmin * 3; j < xmin * 3 + 3; j++)
            if (sudoku[i][j] == value && i != pos.second && j != pos.first)
            {
                return false;
            }
    return true;
}



//find the next empty entry in the board, return it's pos
//return (-1,-1) if the board is already full
std::pair<short, short> find_empty(const std::vector<std::vector<short>>& sudoku) 
{
    for (auto i = sudoku.begin(); i != sudoku.end(); ++i)
        for (auto j = i->begin(); j != i->end(); ++j)
            if (*j == 0)
            {
                //returning j,i coordinates
                return std::make_pair(std::distance(i->begin(), j), std::distance(sudoku.begin(), i));
            }
    return { -1,-1 };
}


//the backtracking algo
void solver(std::vector<std::vector<short>>& sudoku) 
{
    short value;
    //stack to avoid recursion
    std::stack<std::pair<short,short>> last_pos;
    //starting point
    std::pair<short, short> pos = { 0, 0 };
    pos = find_empty(sudoku);
    while (pos.second != -1) 
    {
        //every new value is set to 1
        sudoku[pos.second][pos.first] = 1;
        value = 1;
        while (!check_rules(pos, sudoku)) 
        {
            //if still no solution and val=9, go back in the stack, backtracking...
            if (value == 9) 
            {
                sudoku[pos.second][pos.first] = 0;
                pos = last_pos.top();
                last_pos.pop();
                //increase the old value by one, backtracking...
                value = sudoku[pos.second][pos.first] + 1;
                sudoku[pos.second][pos.first] = value;
                //special case, if the previous val was already 9, go back further
                if (value > 9) 
                {
                    sudoku[pos.second][pos.first] = 0;
                    pos = last_pos.top();
                    last_pos.pop();
                    value = sudoku[pos.second][pos.first] + 1;
                    sudoku[pos.second][pos.first] = value;
                }
            }
            //iterating through 1-9, checking for a solution
            else 
            {
                sudoku[pos.second][pos.first] += 1;
                value = sudoku[pos.second][pos.first];
            }
        }
        //adding the temporary solution to the stack and finding the next 0
        last_pos.push(pos);
        pos = find_empty(sudoku);
    }
    out_sudoku(sudoku);
}


int main(){
    std::vector<std::vector<short>> board = {
        {0,0,0,0,0,0,0,0,1},
        {0,0,0,0,1,0,2,3,0},
        {0,0,8,4,3,0,0,0,7},
        {0,9,6,0,2,7,0,4,8},
        {0,0,0,0,0,9,0,0,0},
        {0,5,0,0,0,0,0,0,0},
        {0,0,4,0,0,5,0,0,0},
        {0,0,0,0,9,0,1,0,0},
        {9,8,2,7,0,0,0,0,0}

    };
    //solver(board);
    out_sudoku(board);
    std::cout << "The solution is:" << std::endl;
    solver(board);
}
